grpc/server/src/backend.cpp

```diff
@@ -5,52 +5,74 @@
 #include <grpc++/security/server_credentials.h>
+#include <mutex>
 
 #include "action/actionrpc_impl.h"
+#include "connection_initiator.h"
 #include "core/corerpc_impl.h"
+#include "dronecore.h"
 #include "log.h"
 #include "mission/missionrpc_impl.h"
 #include "telemetry/telemetryrpc_impl.h"
 
 namespace dronecore {
 namespace backend {
 
-bool DroneCoreBackend::run(const int mavlink_listen_port)
+class DroneCoreBackend::Impl
 {
-    _connection_initiator.start(_dc, 14540);
-    _connection_initiator.wait();
+public:
+    Impl() {}
+    ~Impl() {}
 
-    grpc::ServerBuilder builder;
-    setup_port(builder);
+    bool run(const int mavlink_listen_port)
+    {
+        _connection_initiator.start(_dc, 14540);
+        _connection_initiator.wait();
 
-    CoreServiceImpl core(_dc);
-    builder.RegisterService(&core);
+        grpc::ServerBuilder builder;
+        setup_port(builder);
 
-    Action action(&_dc.device());
-    ActionServiceImpl actionService(action);
-    builder.RegisterService(&actionService);
+        CoreServiceImpl core(_dc);
+        builder.RegisterService(&core);
 
-    Mission mission(&_dc.device());
-    MissionServiceImpl missionService(mission);
-    builder.RegisterService(&missionService);
+        Action action(&_dc.device());
+        ActionServiceImpl actionService(action);
+        builder.RegisterService(&actionService);
 
-    Telemetry telemetry(&_dc.device());
-    TelemetryServiceImpl telemetryService(telemetry);
-    builder.RegisterService(&telemetryService);
+        Mission mission(&_dc.device());
+        MissionServiceImpl missionService(mission);
+        builder.RegisterService(&missionService);
 
-    _server = builder.BuildAndStart();
-    LogInfo() << "Server started";
-    _server->Wait();
+        Telemetry telemetry(&_dc.device());
+        TelemetryServiceImpl telemetryService(telemetry);
+        builder.RegisterService(&telemetryService);
 
-    return true;
-}
+        _server = builder.BuildAndStart();
+        LogInfo() << "Server started";
+        _server->Wait();
 
-void DroneCoreBackend::setup_port(grpc::ServerBuilder &builder)
-{
-    std::string server_address("0.0.0.0:50051");
-    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
-    LogInfo() << "Server set to listen on " << server_address;
-}
+        return true;
+    }
+
+private:
+    void setup_port(grpc::ServerBuilder &builder)
+    {
+        std::string server_address("0.0.0.0:50051");
+        builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
+        LogInfo() << "Server set to listen on " << server_address;
+    }
+
+    dronecore::DroneCore _dc;
+    dronecore::backend::ConnectionInitiator<dronecore::DroneCore> _connection_initiator;
+    std::unique_ptr<grpc::Server> _server;
+};
+
+DroneCoreBackend::DroneCoreBackend() : _impl(new Impl()) {}
+DroneCoreBackend::~DroneCoreBackend() = default;
+DroneCoreBackend::DroneCoreBackend(DroneCoreBackend &&) = default;
```

Comment:
What does `default` mean for these? What about using `delete` ?

Reply:
This time, I think it is explained in the article showing the pImpl coming from Scott Meyers' book ([article here](http://oliora.github.io/2015/12/29/pimpl-and-rule-of-zero.html). Specifically, it says:

> // Tell the compiler to generate default special members which utilize
// the power of std::unique_ptr.
// We can do it here because the implementation class is defined at this point thus
// std::unique_ptr can properly handle the implementation pointer.

Reply:
I can still change it later if you want, but as far as I understand, `default` is not wrong. So I'll already merge that :-).

Reply:
Let's continue this discussion [here](https://github.com/dronecore/DroneCore/pull/300).